## REGULAR EXPRESSION LEARNING
- 1. re.search()          → find a pattern in text
- 2. span                 → index position of match
- 3. []                   → character class — any one character from the set
- 4. [msb]                → match m or s or b
- 5. [A-Z], [a-z], [0-9]  → ranges inside character class
- 6. [Pp]ython            → case insensitive matching
- 7. \d                   → shortcut for [0-9]
- 8. \w                   → shortcut for [A-Za-z0-9_]
- 9. {3}                  → exactly 3 times
- 10. {3,5}               → 3 to 5 times
- 11. {,3}                → 0 to 3 times
- 12. {8,}                → 8 or more times
- 13. *                   → 0 or more (a{0,})
- 14. +                   → 1 or more (a{1,})
- 15. ?                   → 0 or 1  (a{0,1})
- 16. ()                  → capture group — used with .extract()
- 17. .                   → any character except newline
- 18. str.contains()      → filter rows matching pattern
- 19. str.extract()       → pull out matched text
- 20. value_counts()      → count extracted results
  21.  Remove ALL non-digit characters == .str.replace(r"\D", "", regex=True)
 **\D** means "anything that is NOT a digit"
 This removes spaces, dashes, brackets, dots — everything

In [1]:
import pandas as pd
import re

In [2]:
df = pd.read_csv("../Datasets/employees.csv")
df.head(5)

,employee_id,full_name,department,job_title,hire_date,salary,bonus,years_experience,email,phone,city,state,performance_rating,status,remote_work,manager_name,last_review_date
0,1034,Grace Ferreira,IT,Software Engineer,11/08/2021,87609,NaN,10,grace.ferreira@retailco.com,779 898 3950,Nashville,TN,2.0,Active,No,Sophie Khan,4/15/2023
1,1056,Sonia Chen,Operations,Logistics Coordinator,01/21/2019,58508,NaN,4,sonia.chen@retailco.com,(736) 569-2108,Nashville,TN,5.0,Active,No,Hugo Muller,12/01/2024
2,1156,Wei Khan,Finance,Finance Manager,2019-08-02,64418,NaN,1,wei.khan@retailco.com,NaN,Nashville,TN,2.0,Inactive,No,Noah Kowalski,2/23/2024
3,1081,Noah Brown,Operations,Operations Manager,2016-01-07,58431,1560.0,2,noah.brown@retailco.com,(392) 979-4555,Columbus,OH,5.0,Active,Hybrid,NaN,12/25/2023
4,1072,Wei Wilson,OPERATIONS,Operations Manager,8/6/2018,63197,NaN,6,wei.wilson@retailco.com,747-956-1611,Chicago,IL,1.0,On Leave,Hybrid,Ahmed Wilson,01/01/2024


In [3]:
df.columns.to_list()

['employee_id',
 'full_name',
 'department',
 'job_title',
 'hire_date',
 'salary',
 'bonus',
 'years_experience',
 'email',
 'phone',
 'city',
 'state',
 'performance_rating',
 'status',
 'remote_work',
 'manager_name',
 'last_review_date']

In [4]:
email = df["email"]
email

0      grace.ferreira@retailco.com
1          sonia.chen@retailco.com
2            wei.khan@retailco.com
3          noah.brown@retailco.com
4          wei.wilson@retailco.com
                  ...             
219          ravi.lin@retailco.com
220       ivy.fischer@retailco.com
221      sophie.brown@retailco.com
222      mike.popescu@retailco.com
223      sarah.wilson@retailco.com
Name: email, Length: 224, dtype: object

In [5]:
pattern = "\w+"

email[email.str.contains(pattern)]

0      grace.ferreira@retailco.com
1          sonia.chen@retailco.com
2            wei.khan@retailco.com
3          noah.brown@retailco.com
4          wei.wilson@retailco.com
                  ...             
219          ravi.lin@retailco.com
220       ivy.fischer@retailco.com
221      sophie.brown@retailco.com
222      mike.popescu@retailco.com
223      sarah.wilson@retailco.com
Name: email, Length: 224, dtype: object

**Task 1 — Find all job titles containing "Manager"**
- Use str.contains() with simple pattern
- Expected: rows where job_title has Manager or manager

In [6]:
pattern = "[Mm]anager"

job_title = df["job_title"]

job_title[job_title.str.contains(pattern)]

2               Finance Manager
3            Operations Manager
4            Operations Manager
21                Brand Manager
22                Brand Manager
25     Customer Success Manager
33     Customer Success Manager
34              Finance Manager
35              Finance Manager
46              Finance Manager
52     Customer Success Manager
53                Sales Manager
55                Sales Manager
58     Customer Success Manager
59                Sales Manager
61     Customer Success Manager
63              Finance Manager
65     Customer Success Manager
68     Customer Success Manager
69                Brand Manager
79                Sales Manager
80           Operations Manager
81                Sales Manager
87           Operations Manager
93           Operations Manager
99                Brand Manager
102          Operations Manager
103    Customer Success Manager
108             Finance Manager
119               Sales Manager
123               Brand Manager
133     

**Task 2 — Find employees in departments starting with "F" or "I"**

- Use character class []
- Hint: [FI]

In [7]:
pattern = "^[FI]"

# Here, we created new emp that contains 'full_name and department' from df
emp = df[["full_name","department"]]   

# here, we said, in emp, 'department' columns should match pattern
result = emp[emp["department"].str.contains(pattern)]
print(result)

             full_name department
0       Grace Ferreira         IT
8      Miguel Suzuki           IT
10       Lucas Costa          IT 
16         Noah Larsen         IT
32      Victor   Costa   FINANCE 
43     Sophie   Hassan   FINANCE 
66         Peter Costa        IT 
71         Sonia Bello        IT 
73      Sonia   Walker       IT  
74        Peter Chen           IT
77      Simon Robinson   Finance 
78       Mei  Robinson         IT
85         Ravi Turner        IT 
91      Carlos Marquez         IT
104         Tom Dube           IT
112     Wei   Robinson   Finance 
114     Lucas   Nguyen         IT
122       Ali Martin         IT  
132       Jack Alvarez    Finance
142      Maya   Nguyen         IT
144        Leo Johnson    Finance
147       Kenji Khan         IT  
150      Maria Costa    Finance  
151     Layla   Walker   FINANCE 
154         Sam Larsen         IT
158          Sam Rossi        IT 
159      Chloe Popescu        IT 
162       Grace   Dube    FINANCE
165      Lucas

**Task 3 — Find all emails containing numbers in the name part**

- Use \d inside str.contains()
- "grace2.ferreira@retailco.com" → flag it

In [8]:
pattern = "[1-2]\d"

email[email.str.contains(pattern)]

#there is no email that contains number in name part

Series([], Name: email, dtype: object)

In [9]:
email.sample()

101    paula.costa@retailco.com
Name: email, dtype: object

**Task 4 — Find phone numbers that use spaces as separator**

- "779 898 3950" → has spaces
- "(736) 569-2108" → has brackets, not just spaces
- Use \d{3}\s\d{3} to find space-separated format

In [10]:
pattern = r"\d{3}\s\d{3}\s\d{4}"

phone_num = df["phone"]
phone_num.str.contains(pattern)
phone_num[phone_num.str.contains(pattern, na=False)]

0      779 898 3950
16     979 527 4328
23     841 747 3719
24     519 819 6395
27     870 208 2091
32     913 767 9849
34     324 229 4920
35     774 647 7467
42     592 204 1618
44     633 390 8973
50     830 628 5067
57     687 508 8803
60     615 590 8130
65     523 905 1137
68     216 327 3588
70     516 434 5053
74     715 426 1710
75     992 528 5414
78     305 212 5658
79     854 698 6198
81     999 571 9535
84     458 423 8646
95     676 646 5397
97     953 341 2951
111    978 763 1601
112    222 449 1706
118    250 366 4511
119    961 650 7334
120    838 384 8444
122    940 857 9037
143    509 403 7302
151    460 203 4570
155    549 572 1889
167    201 732 9821
170    670 778 4984
171    573 316 4273
175    748 903 8779
178    846 386 3209
180    761 656 3607
183    682 700 5341
184    785 823 8750
186    331 958 5237
190    875 699 1257
191    904 810 3539
196    322 783 5033
199    828 206 9871
204    756 205 9084
205    504 664 2195
208    317 806 8584
209    625 263 4388


In [11]:
# pattern = r"(\d{3}\s\d{3}\s\d{4})| ((\d{3})\s\d{3}-\s\d{4})| (\d{3}-\s\d{3}-\s\d{4})"
pattern = r"(\(\d{3}\)\s\d{3}-\d{4}) | (\d{3}-\d{3}-\d{4}) | (\d{3}\s\d{3}\s\d{4})"

phone_num = df["phone"]
phone_num.str.extract(pattern)
# phone_num[phone_num.str.extract(pattern)]

,0,1,2
0,NaN,NaN,NaN
1,NaN,NaN,NaN
2,NaN,NaN,NaN
3,NaN,NaN,NaN
4,NaN,NaN,NaN
...,...,...,...
219,NaN,NaN,NaN
220,NaN,NaN,NaN
221,NaN,NaN,NaN
222,NaN,NaN,NaN


**Task 5 — Extract area code from phone numbers**

- "779 898 3950"   → "779"
- "(736) 569-2108" → "736"
- "747-956-1611"   → "747"
- Hint: use \d{3} with capture group ()

In [206]:
# pattern = r"(\d{3})\s|\((\d{3})\)|(\d{3})-"

pattern = r"(\d{3})"
result = phone_num.str.extract(pattern)

print(result)


# More precise — handle all 3 formats:
# pattern = r"\(?(\d{3})\)?"
# result = phone_num.str.extract(pattern)
# print(result)


       0
0    779
1    736
2    NaN
3    392
4    747
..   ...
219  977
220  292
221  906
222  894
223  840

[224 rows x 1 columns]


**Task 6 — Find all employees whose full_name has exactly two words**

- "Grace Ferreira" → two words ✅
- "Mary Jane Watson" → three words ❌
- Hint: use \w+ for each word, \s for space

In [212]:
name = df["full_name"]
name

# Correct pattern:
# pattern = r"^\w+\s\w+$"

pattern = r"\w+\s\w"

name[name.str.contains(pattern)]




0         Grace Ferreira
2               Wei Khan
3             Noah Brown
8        Miguel Suzuki  
10         Lucas Costa  
             ...        
214      Kenji Andersson
215         Diego Ruiz  
216          Lena Osei  
220        Ivy Fischer  
223         Sarah Wilson
Name: full_name, Length: 136, dtype: object

**Task 7 — Extract year from hire_date**

- "11/08/2021" → "2021"
- "2019-08-02" → "2019"
- "8/6/2018"   → "2018"

In [14]:
hiring = df["hire_date"]
hiring

pattern = r"(\d{4})"

result = hiring.str.extract(pattern)

print(result)

        0
0    2021
1    2019
2    2019
3    2016
4    2018
..    ...
219  2018
220  2018
221  2014
222  2019
223  2016

[224 rows x 1 columns]


**Task 8 — Find phone numbers that have exactly 10 digits**

- First remove all non-digit characters
- Then check if what remains is exactly 10 digits

In [15]:
# cleaning non digits character using == \D

num_clean = phone_num.str.replace(r"\D", "", regex=True)

# finding 10 digits phone numbers
pattern = r"\d{10}$"
result = phone_num[num_clean.str.match(pattern, na=False)]

print(result)

0        779 898 3950
1      (736) 569-2108
3      (392) 979-4555
4        747-956-1611
5        340.311.4920
            ...      
219    (977) 873-7164
220      292 939 9202
221      906 689 7909
222      894.973.3240
223      840.803.4241
Name: phone, Length: 203, dtype: object


**Task 9 — Extract first and last name separately using capture groups**

- "Grace Ferreira" → Group 1: "Grace", Group 2: "Ferreira"

In [67]:
email

pattern = r"(\w+)\.(\w+)"
result = email.str.extract(pattern)

print(result)

          0         1
0     grace  ferreira
1     sonia      chen
2       wei      khan
3      noah     brown
4       wei    wilson
..      ...       ...
219    ravi       lin
220     ivy   fischer
221  sophie     brown
222    mike   popescu
223   sarah    wilson

[224 rows x 2 columns]


**Task 10 — Find job titles that contain Senior or Junior or Lead**

- Use [] or | — but since you haven't learned | yet, use what you know
- Hint: search separately for each one
- `[SJ]` won't work here — think about how to approach this differently

In [213]:
job_title

pattern = r"[Ss]enior|[Jj]unior|[Ll]ead"

job_title[job_title.str.contains(pattern, na=False)]

13     Regional Sales Lead
70     Regional Sales Lead
181    Regional Sales Lead
192    Regional Sales Lead
202    Regional Sales Lead
Name: job_title, dtype: object